### PR TITLE
feat(sequencer)!: implement bridge accounts and related actions

### PR DESCRIPTION
**Pull request by noot**
## Summary
implement bridge accounts and related actions (`InitBridgeAccountAction` and `BridgeLockAction`).

at a high level, the flow for bridging is:
- a rollup decides to initialize a bridge on the sequencer by sending a `InitBridgeAccountAction` which registers a sequencer address to a specific rollup ID; this address becomes the address of the bridge, ie. the owner who is authorized to withdraw from the account. right now it's just a normal account, but could potentially be generated using threshold signatures off-chain
- a user can bridge to the rollup by sending funds to a bridge account via `BridgeLockAction`, which is a normal transfer but also specifies a `destination_chain_address`, which represents an address on the rollup to which the bridged funds will be sent.
- when a `BridgeLockAction` is executed, a `Deposit` is "emitted" by the sequencer during `EndBlock` (this is unused right now but will be written to DA and returned by the sequencer API as part of a rollup's data in a later PR)
- the bridge account can withdraw funds from itself by sending a `TransferAction` from their own account to another account.

## Background
we want the ability to potentially bridge tokens from the sequencer to a rollup on it and back.

## Changes
- implement `InitBridgeAccountAction` for initializing a bridge account
- implement `BridgeLockAction` for transferring to the bridge account
- add a `Deposit` type which is what will be part of the rollup data indicating a bridge lock has happened
- small improvements like adding `increase_balance` and `decrease_balance` state methods

## Testing
unit tests

## Breaking Changelist
- two new actions are added: `InitBridgeAccountAction` and `BridgeLockAction`
- a new bridge account type is added (changes to state)

### CODE REVIEW

File: proto/sequencerapis/astria/sequencer/v1alpha1/transaction.proto [hunk 1]
```diff
@@ -127,3 +129,32 @@ message FeeAssetChangeAction {
...
 }
+
+// `InitBridgeAccountAction` represents a transaction that initializes
+// a bridge account for the given rollup on the chain.
+//
+// The sender of the transaction is used as the owner of the bridge account
+// and is the only actor authorized to transfer out of this account via
+// a `TransferAction`.
+message InitBridgeAccountAction {
+  bytes rollup_id = 1;
```
Comment (SuperFluffy) [inline proto/sequencerapis/astria/sequencer/v1alpha1/transaction.proto:140]: Can you add comments on these fields? Specifically constraints if there are any?

File: proto/sequencerapis/astria/sequencer/v1alpha1/data.proto
Comment (SuperFluffy) [inline proto/sequencerapis/astria/sequencer/v1alpha1/data.proto:1]: Nuking this because it's not used, right?

File: proto/sequencerapis/astria/sequencer/v1alpha1/block.proto [hunk 2]
```diff
@@ -41,3 +42,19 @@ message SequencerBlock {
   // the rollup transactions.
   astria.sequencer.v1alpha1.Proof rollup_ids_proof = 4;
 }
+
+// `Deposit` represents a deposit from the sequencer
```
Comment (SuperFluffy) [inline proto/sequencerapis/astria/sequencer/v1alpha1/block.proto:46]: Maybe?
```suggestion
// `Deposit` is a transfer of funds from the sequencer to a rollup.
```
Also, in which context is this used? Can you explain why this is not an action? From a first glance it looks very action-y.

File: proto/sequencerapis/astria/sequencer/v1alpha1/block.proto [hunk 3]
```diff
@@ -41,3 +42,19 @@ message SequencerBlock {
...
+// to a rollup.
+message Deposit {
+  // the address the funds were locked in on the sequencer.
+  // this is required as initializing an account as a bridge account
+  // is permissionless, so the rollup consensus needs to know and enshrine
+  // which accounts it accepts as valid bridge accounts.
+  bytes bridge_address = 1;
+  bytes rollup_id = 2;
```
Comment (SuperFluffy) [inline proto/sequencerapis/astria/sequencer/v1alpha1/block.proto:54]: Can you add explainers to all of these fields that currently lack them?

File: crates/astria-sequencer/src/accounts/state_ext.rs [hunk 4]
```diff
@@ -157,6 +158,48 @@ pub(crate) trait StateWriteExt: StateWrite {
...
+        let balance = self
+            .get_account_balance(address, asset)
+            .await
+            .context("failed to get account balance")?;
+        self.put_account_balance(
+            address,
+            asset,
+            balance
+                .checked_sub(amount)
+                .ok_or(anyhow!("account balance underflow; insufficient funds?"))?,
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/accounts/state_ext.rs:199]: `ok_or(anyhow!(...))` is not necessary for anyhow, as it impls `Context for Option<T>`:

```suggestion
                .context("subtracting from account balance failed due to insufficient funds")?,
```

File: crates/astria-sequencer/src/accounts/state_ext.rs [hunk 5]
```diff
@@ -157,6 +158,48 @@ pub(crate) trait StateWriteExt: StateWrite {
...
+        )?;
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/accounts/state_ext.rs:200]: context

File: crates/astria-sequencer/src/accounts/state_ext.rs [hunk 6]
```diff
@@ -157,6 +158,48 @@ pub(crate) trait StateWriteExt: StateWrite {
...
+        let balance = self
+            .get_account_balance(address, asset)
+            .await
+            .context("failed to get account balance")?;
+        self.put_account_balance(
+            address,
+            asset,
+            balance
+                .checked_add(amount)
+                .ok_or(anyhow!("account balance overflow"))?,
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/accounts/state_ext.rs:178]: Same as for the decrease_balance bit below, anyhow can be used directly on options:
```suggetion
                .context("adding to account balance failed due to overflow")?,
```

File: crates/astria-sequencer/src/accounts/state_ext.rs [hunk 7]
```diff
@@ -157,6 +158,48 @@ pub(crate) trait StateWriteExt: StateWrite {
...
+        )?;
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/accounts/state_ext.rs:179]: context, i.e.:
```suggestion
        ).context("failed storing updated account balance in database"?;
```

File: crates/astria-sequencer/src/accounts/action.rs [hunk 8]
```diff
@@ -25,56 +26,74 @@ use crate::{
...
         from: Address,
     ) -> Result<()> {
+        // ensure the recipient is not a bridge account,
+        // as the explicit `BridgeLockAction` should be used to transfer to a bridge account.
         ensure!(
-            state.is_allowed_fee_asset(self.fee_asset_id).await?,
-            "invalid fee asset",
+            state
+                .get_bridge_account_rollup_id(&self.to)
+                .await?
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/accounts/action.rs:91]: context

File: crates/astria-sequencer/src/accounts/action.rs [hunk 9]
```diff
@@ -25,56 +26,74 @@ use crate::{
...
-                .await
-                .context("failed to get account balance in transfer check")?;
-            ensure!(
-                from_transfer_balance >= self.amount,
-                "insufficient funds for transfer"
-            );
-        }
-
-        Ok(())
+        transfer_check_stateful(self, state, from).await
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/accounts/action.rs:96]: context since there are more than this error in function

File: crates/astria-sequencer/src/accounts/action.rs [hunk 10]
```diff
@@ -25,56 +26,74 @@ use crate::{
...
+        );
+    } else {
+        // otherwise, check the fee asset account has enough to cover the fees,
+        // and the transfer asset account has enough to cover the transfer
+        ensure!(
+            from_fee_balance >= TRANSFER_FEE,
+            "insufficient funds for fee payment"
+        );
+
+        let from_transfer_balance = state
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/accounts/action.rs:66]: Nit: the naming here tripped me up. What about calling this just `from_balance`?

File: crates/astria-sequencer/src/accounts/action.rs [hunk 11]
```diff
@@ -25,56 +26,74 @@ use crate::{
...
+        .await
+        .context("failed getting `from` account balance for fee payment")?;
+
+    // if fee asset is same as transfer asset, ensure accounts has enough funds
+    // to cover both the fee and the amount transferred
+    if action.fee_asset_id == transfer_asset_id {
+        let payment_amount = action
+            .amount
+            .checked_add(TRANSFER_FEE)
+            .ok_or(anyhow!("transfer amount plus fee overflowed"))?;
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/accounts/action.rs:52]: no need to `ok_or(anyhow!`:

```suggestion
            .context("transfer amount plus fee overflowed")?;
```

File: crates/astria-sequencer/src/accounts/action.rs [hunk 12]
```diff
@@ -25,56 +26,74 @@ use crate::{
 /// Fee charged for a `Transfer` action.
 pub(crate) const TRANSFER_FEE: u128 = 12;
 
+pub(crate) async fn transfer_check_stateful<S: StateReadExt + 'static>(
+    action: &TransferAction,
+    state: &S,
+    from: Address,
+) -> Result<()> {
+    ensure!(
+        state.is_allowed_fee_asset(action.fee_asset_id).await?,
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/accounts/action.rs:35]: context

File: crates/astria-core/src/sequencer/v1alpha1/transaction/action.rs [hunk 13]
```diff
@@ -975,3 +1012,198 @@ enum FeeAssetChangeActionErrorKind {
...
+    fn invalid_fee_asset_id(err: asset::IncorrectAssetIdLength) -> Self {
+        Self(BridgeLockActionErrorKind::InvalidFeeAssetId(err))
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+enum BridgeLockActionErrorKind {
+    #[error("the address was invalid")]
+    InvalidAddress(#[source] IncorrectAddressLength),
+    #[error("the amount was missing")]
```
Comment (SuperFluffy) [inline crates/astria-core/src/sequencer/v1alpha1/transaction/action.rs:1203]:
```suggestion
    #[error("the `amount` field was not set")]
```
I think this makes it a bit more clear what the issue is.

File: crates/astria-core/src/sequencer/v1alpha1/transaction/action.rs [hunk 14]
```diff
@@ -975,3 +1012,198 @@ enum FeeAssetChangeActionErrorKind {
...
+    }
+
+    #[must_use]
+    fn invalid_fee_asset_id(err: asset::IncorrectAssetIdLength) -> Self {
+        Self(BridgeLockActionErrorKind::InvalidFeeAssetId(err))
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+enum BridgeLockActionErrorKind {
```
Comment (SuperFluffy) [inline crates/astria-core/src/sequencer/v1alpha1/transaction/action.rs:1200]: Can you highlight the bad field names, e.g. "the `amount` field was not set`" or "the `asset_id` field was invalid"?

File: crates/astria-core/src/sequencer/v1alpha1/transaction/action.rs [hunk 15]
```diff
@@ -975,3 +1012,198 @@ enum FeeAssetChangeActionErrorKind {
...
+    }
+
+    #[must_use]
+    fn invalid_fee_asset_id(err: asset::IncorrectAssetIdLength) -> Self {
+        Self(InitBridgeAccountActionErrorKind::InvalidFeeAssetId(err))
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+#[allow(clippy::enum_variant_names)]
```
Comment (SuperFluffy) [inline crates/astria-core/src/sequencer/v1alpha1/transaction/action.rs:1100]: I think the `allow` here is unnecessary. Just remove the `Invalid*` prefixes.

In general, please add explainers why we deviate from the pedantic clippy. It's ok for the `*Action` types because that's what the entire crate does (to match the protos), but otherwise I wouldn't do it

File: crates/astria-core/src/sequencer/v1alpha1/block.rs [hunk 16]
```diff
@@ -565,3 +568,77 @@ where
...
+            rollup_id,
+            amount,
+            asset_id,
+            destination_chain_address,
+        })
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum DepositError {
```
Comment (SuperFluffy) [inline crates/astria-core/src/sequencer/v1alpha1/block.rs:635]: This library tries to not make the inner errors part of its public interface. Please make this a newtype wrapper: `pub struct DepositError(DepositErrorKind)` and then define a private `enum DepositErrorKind`

File: crates/astria-core/src/sequencer/v1alpha1/block.rs [hunk 17]
```diff
@@ -565,3 +568,77 @@ where
...
+    #[error(transparent)]
```
Comment (SuperFluffy) [inline crates/astria-core/src/sequencer/v1alpha1/block.rs:636]: Please don't make this or the other error variants transparent.

File: crates/astria-core/src/sequencer/v1alpha1/asset.rs [hunk 18]
```diff
@@ -157,6 +157,12 @@ impl From<String> for Id {
     }
 }
 
+impl From<[u8; 32]> for Id {
```
Comment (SuperFluffy) [inline crates/astria-core/src/sequencer/v1alpha1/asset.rs:160]: Is this used anywhere?

File: crates/astria-core/src/sequencer/v1alpha1/block.rs [hunk 19]
```diff
@@ -565,3 +568,77 @@ where
     };
     merkle::Tree::from_leaves(iter.into_iter().map(|item| Sha256::digest(&item)))
 }
+
+/// [`Deposit`] represents a deposit from the sequencer to a rollup.
+///
+/// A [`Deposit`] is constructed whenever a [`BridgeLockAction`] is executed
+/// and stored as part of the block's events.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Deposit {
```
Comment (SuperFluffy) [inline crates/astria-core/src/sequencer/v1alpha1/block.rs:577]: This crate tries to keep most fields private by default while this type doesn't. I grepped for a few fields but couldn't quickly find direct field accesses.

Can we keep these fields private?

**Review by SuperFluffy — COMMENTED** [review]
Bunch of style things, missing error context and consistency comments

File: proto/sequencerapis/astria/sequencer/v1alpha1/data.proto
Comment (noot) [reply proto/sequencerapis/astria/sequencer/v1alpha1/data.proto:1]: yes exactly

File: proto/sequencerapis/astria/sequencer/v1alpha1/block.proto (hunk 2 above)
Comment (noot) [reply proto/sequencerapis/astria/sequencer/v1alpha1/block.proto:46]: it's more similar to an "event", in that the sequencer application detects deposits and their corresponding rollup IDs so that the conductors/rollups don't have parse all the bridge lock txs in a block. a `Deposit` isn't exactly a transfer of funds (since that's implemented by the rollup), it's just notifying a rollup that funds were locked to some account on the sequencer. it's up to the rollup what to do with that info

Comment (noot) [reply proto/sequencerapis/astria/sequencer/v1alpha1/block.proto:46]: will add this to the comment

File: crates/astria-sequencer/src/accounts/action.rs (hunk 10 above)
Comment (noot) [reply crates/astria-sequencer/src/accounts/action.rs:66]: there's two from balances though, this is to distinguish it from the `from_fee_balance`

File: crates/astria-core/src/sequencer/v1alpha1/transaction/action.rs (hunk 15 above)
Comment (noot) [reply crates/astria-core/src/sequencer/v1alpha1/transaction/action.rs:1100]: i put the `Invalid` prefix for consistency for other errors, just removed it + the clippy allow

File: crates/astria-core/src/sequencer/v1alpha1/asset.rs (hunk 18 above)
Comment (noot) [reply crates/astria-core/src/sequencer/v1alpha1/asset.rs:160]: yes, in `astria-sequencer/src/bridge/state_ext.rs:103`: `asset_ids.0.into_iter().map(asset::Id::from).collect()`

File: crates/astria-core/src/sequencer/v1alpha1/transaction/action.rs (hunk 15 above)
Comment (noot) [reply crates/astria-core/src/sequencer/v1alpha1/transaction/action.rs:1103]: ok nvm, reverted it back and added a comment. clippy is angry still even if i remove the prefix: `error: all variants have the same postfix: `Id`` but removing the postfix makes no sense

File: crates/astria-core/src/sequencer/v1alpha1/block.rs [hunk 20]
```diff
@@ -565,3 +568,144 @@ where
...
+    bridge_address: Address,
+    rollup_id: RollupId,
+    amount: u128,
+    asset_id: asset::Id,
+    destination_chain_address: String,
```
Comment (joroshiba) [inline crates/astria-core/src/sequencer/v1alpha1/block.rs:582]: nit: comments on field values

File: crates/astria-core/src/sequencer/v1alpha1/transaction/action.rs [hunk 21]
```diff
@@ -975,3 +1012,201 @@ enum FeeAssetChangeActionErrorKind {
     #[error("the asset_id was missing")]
     MissingAssetId,
 }
+
+#[allow(clippy::module_name_repetitions)]
+#[derive(Debug, Clone)]
+pub struct InitBridgeAccountAction {
```
Comment (joroshiba) [inline crates/astria-core/src/sequencer/v1alpha1/transaction/action.rs:1018]: nit: comments would be good on struct type and fields.

File: crates/astria-sequencer/src/accounts/action.rs [hunk 22]
```diff
@@ -111,59 +126,30 @@ impl ActionHandler for TransferAction {
...
             state
-                .put_account_balance(
-                    from,
-                    self.fee_asset_id,
-                    from_fee_balance
-                        .checked_sub(TRANSFER_FEE)
-                        .ok_or(anyhow!("insufficient funds for fee payment"))?,
-                )
-                .context("failed updating `from` account balance for fee payment")?;
+                .decrease_balance(from, self.fee_asset_id, TRANSFER_FEE)
```
Comment (joroshiba) [inline crates/astria-sequencer/src/accounts/action.rs:150]: decrease balance is a lot cleaner, nice.

File: proto/sequencerapis/astria/sequencer/v1alpha1/data.proto
Comment (joroshiba) [inline proto/sequencerapis/astria/sequencer/v1alpha1/data.proto:1]: Why is this removed here?

**Review by joroshiba — COMMENTED** [review]
Generally looks good, couple nits, tests are failing due to updated access fields

File: proto/sequencerapis/astria/sequencer/v1alpha1/data.proto
Comment (noot) [reply proto/sequencerapis/astria/sequencer/v1alpha1/data.proto:1]: it wasn't used anywhere

Comment (joroshiba) [inline proto/sequencerapis/astria/sequencer/v1alpha1/data.proto:1]: nit: unrelated change. Leave for future cleanup

**Review by joroshiba — APPROVED** [review]
LGTM, outside failing tests, minor nit wrt deleting proto file not in use.

Should we consider slapping a deprecated tag on `put_account_balance` and remove in favor of the explicit increase/decrease methods? This can be a follow on task.

File: proto/sequencerapis/astria/sequencer/v1alpha1/data.proto
Comment (noot) [reply proto/sequencerapis/astria/sequencer/v1alpha1/data.proto:1]: should i make a separate PR to delete it?? i want it gone since it's confusing

Comment (noot) [reply proto/sequencerapis/astria/sequencer/v1alpha1/data.proto:1]: https://github.com/astriaorg/astria/pull/783
